Clone Activity tags to Span Attributes

brunomasantos opened the pull request:
Draft work to solve the issue #430 . Has its my first contribution would like to understand if this is the way to go or not.

Didn't have concerns with code optimization, just wanted to make sure this was the correct approach so that i can refine it. 

Thanks

brunomasantos commented:
I signed it

tarekgh commented on src/OpenTelemetry/Trace/SpanSdk.cs:
```diff
@@ -615,13 +615,18 @@ public void Dispose()
         {
+            var spanCreationOptions = new SpanCreationOptions
+            {
+                Attributes = activity.Tags?.ToDictionary(k => k.Key, v => (object)v.Value),
```
activity.Tags always return non-null value. It would be better to test the count instead. 

```C$
var tags = activity.Tags;
if (tags.Count() > 0)
{
....
}
```

tarekgh replied:
@SergeyKanzhelev why we have SpanCreationOptions.Attributes defined as IDictionary<string, object> instead of `IEnumerable<KeyValuePait<string, object>>`? I am asking because we can avoid allocating the intermediate IDictionary object and pass IEnumerable directly which we got from Activity.Tags.

brunomasantos replied:
Thanks for the quick review, will make the changes later then.

About the dictionary I also had that question, wondering if it is to guarantee the uniqueness of the keys?

tarekgh replied:
> wondering if it is to guarantee the uniqueness of the keys?

Yes, the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-attributes) is saying the keys should be unique. but at the same time, it requires keeping the order of the keys too which the Dictionary doesn't guarantee that.

```
Attributes SHOULD preserve the order in which they're set. Setting an attribute with the same key as an existing attribute SHOULD overwrite the existing attribute's value.
```

brunomasantos replied:
Yes thats true, and after checking some parts of the code dont see any specific reason to be a dictionary. Like you stated the keys are being overwriten when they are repeated and the `IEnumerable` would avoid unecessary memory allocation in cases like this PR.

Also I made the changes that you mentioned

SergeyKanzhelev commented:
@brunomasantos please sign CLA. Also go ahead and remove Draft flag =)

This PR looks OK. I think we can track a separate item on converting `Dictionary` to `IEnumerable`. @tarekgh would you mind creating it?

tarekgh commented:
I have created https://github.com/open-telemetry/opentelemetry-dotnet/issues/588

MikeGoldsmith commented:
I've updated the PR ready for review and updated the branch based on master. @brunomasantos please can you check you've signed CLA.

MikeGoldsmith commented:
Can you also update the PR title - thanks!

brunomasantos commented:
> I've updated the PR ready for review and updated the branch based on master. @brunomasantos please can you check you've signed CLA.

I actually have sign it since the beginning, I am trying to understand what is the problem
